### PR TITLE
Shading rate dynamic

### LINKED ISSUE #341: Fragment shading rate loses device on resize
Resizing the window on the Fragment shading rate causes a lost device on next frame draw.

### PR DESCRIPTION
**Pull request by gpx1000**
## Description
Render a simple scene showing the basics of shading rate dynamic.  This demo shows low and high frequency textures over several cubes.  It creates a sample rate map based upon this frequency every frame, then it uses that dynamic sample rate map as a base for the next frame.

## Tested Environments
Windows 10 / Ubuntu Linux
GeForce 3060 / GeForce 2080 / GeForce 1070
Vulkan version: 1.2.176
Sample Assets changes
depends upon flame texture found in https://github.com/KhronosGroup/Vulkan-Samples-Assets/pull/14

NB: additionally Fixes: #341 

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

**Comment by gpx1000** [comment]
NB: This also fixes #341

File: samples/extensions/fragment_shading_rate_dynamic/fragment_shading_rate_dynamic.cpp [hunk 1]
```diff
@@ -0,0 +1,909 @@
...
+
+	submit_info.commandBufferCount   = 1;
+	submit_info.pCommandBuffers      = &draw_cmd_buffers[current_buffer];
+	submit_info.signalSemaphoreCount = 2;
+	submit_info.pSignalSemaphores    = semaphores.data();
+
+	VK_CHECK(vkQueueSubmit(queue, 1, &submit_info, VK_NULL_HANDLE));
+	ApiVulkanSample::submit_frame();
+
+	const VkPipelineStageFlags wait_mask           = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
```
Comment (cforfang) [inline samples/extensions/fragment_shading_rate_dynamic/fragment_shading_rate_dynamic.cpp]: `VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT` -- we want to block COMPUTE until the semaphore signals.

File: samples/extensions/fragment_shading_rate_dynamic/fragment_shading_rate_dynamic.cpp [hunk 2]
```diff
@@ -0,0 +1,909 @@
...
+	image_copy.dstOffset      = {0, 0, 0};
+	image_copy.dstSubresource = {VK_IMAGE_ASPECT_COLOR_BIT, 0, 0, 1};
+	image_copy.extent         = shading_rate_image->get_extent();
+	image_copy.srcOffset      = {0, 0, 0};
+
+	VkImageSubresourceRange subresource_range = {VK_IMAGE_ASPECT_COLOR_BIT, 0, 1, 0, 1};
+	vkb::set_image_layout(command_buffer, shading_rate_image_compute->get_handle(), VK_IMAGE_LAYOUT_GENERAL, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, subresource_range);
+	vkb::set_image_layout(command_buffer, shading_rate_image->get_handle(), VK_IMAGE_LAYOUT_FRAGMENT_SHADING_RATE_ATTACHMENT_OPTIMAL_KHR, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, subresource_range);
+
+	vkCmdCopyImage(command_buffer, shading_rate_image_compute->get_handle(), VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, shading_rate_image->get_handle(), VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1, &image_copy);
```
Comment (cforfang) [inline samples/extensions/fragment_shading_rate_dynamic/fragment_shading_rate_dynamic.cpp]: Hm, I'm not sure the above layout transitions takes care of making the image writes by compute visible to transfer. In `set_image_layout()` there's no handling of `old_layout = GENERAL` currently -- see the switch statement in that function. Could either edit that or -- my preference -- make the sync explicit in the code here. :) 

Possibly one should also add handling of `src_layout = VK_IMAGE_LAYOUT_FRAGMENT_SHADING_RATE_ATTACHMENT_OPTIMAL_KHR` as well to make the transfer to the rate image available to fragment?

File: samples/extensions/fragment_shading_rate_dynamic/fragment_shading_rate_dynamic.cpp (hunk 1 above)
Comment (gpx1000) [reply samples/extensions/fragment_shading_rate_dynamic/fragment_shading_rate_dynamic.cpp]: nice catch thanks!

File: samples/extensions/fragment_shading_rate_dynamic/fragment_shading_rate_dynamic.cpp (hunk 2 above)
Comment (gpx1000) [reply samples/extensions/fragment_shading_rate_dynamic/fragment_shading_rate_dynamic.cpp]: We went with some memory barriers as I think that makes the most sense in this case.

**Comment by SaschaWillems** [comment]
The sample looks good in my opinion. One thing I noticed though, is some flicker when running the sample in the default window size:

https://user-images.githubusercontent.com/10283127/134958251-83157dd3-c474-4f41-9220-4417b10f6fa1.mp4

The flicker stops when increasing the window size. I also think the texture should be flipped ;)

What I'd like to see is some additional documentation/comments on how the shading rate is influenced by the frequency, e.g. in the compute shader.

**Review by gary-sweet — COMMENTED** [review]
Correctly reports as unsupported for me, so I'm obviously unable to test or approve.

**Comment by gpx1000** [comment]
@SaschaWillems I believe the issues are fixed and the latest version is up.  Thanks for the review earlier.

**Review by SaschaWillems — APPROVED** [review]
I can confirm that the flickering has been fixed.

Sample is looking great, esp. the different visualization options :)

File: shaders/fragment_shading_rate_dynamic/scene.frag [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		{
+			color   = texture(samplerEnvMap, vec2(inUV.s, 1.0 - inUV.t));
+			vec3 dx = dFdx(color.xyz);
+			vec3 dy = dFdx(color.xyz);
+			freq_x  = dot(dx, dx);
+			freq_y  = dot(dy, dy);
+		}
+		break;
+
+		case 1:        // Phong shading
```
Comment (TomAtkinsonArm) [inline shaders/fragment_shading_rate_dynamic/scene.frag:69]: Can we use shader definitions here?

**Review by TomAtkinsonArm — APPROVED** [review]
Will not have hardware with the supported extension until next week

Sample looks good and reads well. One comment on the use of shader variants over push constants but I assume this is do to the fact that the ApiSamples don't use the frameworks shader variants?

Happy to merge as is

**Review by mklefrancois — APPROVED** [review]
Tested and works well (GeForce RTX 3060 Ti)

**Comment by marty-johnson59** [comment]
Merging with 3 approvals

**Pull request merged**